### PR TITLE
[WIP] [DSW] Preparations for smart contracts

### PR DESCRIPTION
**Pull request by bedri**
This pull will serve to the preparation of the code to the smart contracts

### CODE REVIEW

File: src/script/standard.cpp [hunk 1]
```diff
@@ -327,6 +327,128 @@ CScript GetScriptForMultisig(int nRequired, const std::vector<CPubKey>& keys)
     return script;
 }
 
+/*
```
Comment (pedro-at-decenomy) [inline src/script/standard.cpp:330]: Do we need now all these methods? 
I think that it is better to implement a more generic and flexible solution.

**Review by crypt0p3t3r — COMMENTED** [review]
Instead of creating a folder specifically called smartcontract we should integrate the more capable script capabilities into the script folder, for example:
- to store the published scripts we should have a scriptdb.h/cpp files, where their persistence will be stored;
- to store the state of each script we should also use the same scriptdb.h/cpp files;

File: src/script/script.cpp [hunk 2]
```diff
@@ -15,144 +15,157 @@ const char* GetOpName(opcodetype opcode)
...
+	case OP_NOP3                   : return "OP_NOP3";                  // OP_NOP3
+	case OP_NOP4                   : return "OP_NOP4";                  // OP_NOP4
+	case OP_NOP5                   : return "OP_NOP5";                  // OP_NOP5
+	case OP_NOP6                   : return "OP_NOP6";                  // OP_NOP6
+	case OP_NOP7                   : return "OP_NOP7";                  // OP_NOP7
+	case OP_NOP8                   : return "OP_NOP8";                  // OP_NOP8
+	case OP_NOP9                   : return "OP_NOP9";                  // OP_NOP9
+	case OP_NOP10                  : return "OP_NOP10";                 // OP_NOP10
+
+	// zerocoin
```
Comment (pedro-at-decenomy) [inline src/script/script.cpp:144]: These ones need also to be removed/disabled, however on another branch to remove all the zero coin stuff.

File: src/script/script.cpp [hunk 3]
```diff
@@ -15,144 +15,157 @@ const char* GetOpName(opcodetype opcode)
...
+	case OP_ZEROCOINMINT           : return "OP_ZEROCOINMINT";
+	case OP_ZEROCOINSPEND          : return "OP_ZEROCOINSPEND";
+	case OP_ZEROCOINPUBLICSPEND    : return "OP_ZEROCOINPUBLICSPEND";
+
+	// memory store
+	case OP_MLOAD				   : return "OP_MLOAD";
+	case OP_MSTORE				   : return "OP_MSTORE";
+	case OP_MSTORE8				   : return "OP_MSTORE8";
```
Comment (pedro-at-decenomy) [inline src/script/script.cpp:152]: lets do only the OP_MSTORE for now, we add the 8 bit version later if we really need it

File: src/script/script.cpp [hunk 4]
```diff
@@ -15,144 +15,157 @@ const char* GetOpName(opcodetype opcode)
...
+
+	// storage
+	case OP_SLOAD				   : return "OP_SLOAD";
+	case OP_SSTORE				   : return "OP_SSTORE";
+
+	// smart contract actions
+	case OP_SC_PUBLISH			   : return "OP_SC_PUBLISH";
+	case OP_SC_RUN			   	   : return "OP_SC_RUN";
```
Comment (pedro-at-decenomy) [inline src/script/script.cpp:160]: Why not only OP_PUBLISH and OP_RUN? SC seems redundant

File: src/script/script.cpp [hunk 5]
```diff
@@ -15,144 +15,157 @@ const char* GetOpName(opcodetype opcode)
...
+
+	// cold staking
+	case OP_CHECKCOLDSTAKEVERIFY   : return "OP_CHECKCOLDSTAKEVERIFY";
```
Comment (pedro-at-decenomy) [inline src/script/script.cpp:163]: This one will be remove on other branch

File: src/script/interpreter.cpp [hunk 6]
```diff
@@ -969,6 +969,19 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                 }
                 break;
 
+				// memory store
```
Comment (pedro-at-decenomy) [inline src/script/interpreter.cpp:972]: the indentation is way off

File: src/script/standard.cpp (hunk 1 above)
Comment (bedri) [reply src/script/standard.cpp:330]: > Do we need now all these methods?
> I think that it is better to implement a more generic and flexible solution.

Some of them may have a future use but I think these scripts all have a use case while working with shopping applications or likewise. A more generic one will be very much complex to implement RPC functions on them. `escrow` should have it's own RPC call like `createmultisig`.
I actually thought that it won't hurt to have them now and when we put a generic function there, we can remove the replaced ones.

**Comment by bedri** [comment]
> Instead of creating a folder specifically called smartcontract we should integrate the more capable script capabilities into the script folder, for example:
> 
> * to store the published scripts we should have a scriptdb.h/cpp files, where their persistence will be stored;
> * to store the state of each script we should also use the same scriptdb.h/cpp files;

Ok. It's logical.

File: src/script/interpreter.cpp (hunk 6 above)
Comment (bedri) [reply src/script/interpreter.cpp:972]: I don't know why github shows it like this. In VSCode it seems ok

image

File: src/script/script.cpp (hunk 3 above)
Comment (bedri) [reply src/script/script.cpp:152]: :+1:

File: src/script/script.cpp (hunk 4 above)
Comment (bedri) [reply src/script/script.cpp:160]: At first, I do it like OP_PUBLISH and OP_RUN. Then I thought that in the future we may need OP_RUN for another coding so decided to branch them out as OP_SC_PUBLISH and OP_SC_RUN.
OP_PUBLISH may be used in Proof of Ownership as well (to publish data into a database) in the future for instance.

But it's ok. Let's make it as OP_PUBLISH and OP_RUN and if we will need other PUBLISH or RUN, we can branch out them in the future like OP_OWNERSHIP_PUBLISH etc.

File: src/script/script.cpp (hunk 5 above)
Comment (bedri) [reply src/script/script.cpp:162]: :+1:

File: src/script/script.cpp (hunk 2 above)
Comment (bedri) [reply src/script/script.cpp:144]: :+1:

File: src/script/standard.cpp (hunk 1 above)
Comment (pedro-at-decenomy) [reply src/script/standard.cpp:330]: yes, but it is better to add them afterward and if it justifies their inclusion
if this development works well, at the end there will not be the need to have those very or too specific methods

File: src/script/interpreter.cpp (hunk 6 above)
Comment (pedro-at-decenomy) [reply src/script/interpreter.cpp:973]: it seems the classical problem between tabs and spaces

try to use these setting on vscode:
- Editor: Detect Indentation
- Editor: Tab Size 4
- Editor: Insert Spaces

File: src/script/standard.cpp (hunk 1 above)
Comment (bedri) [reply src/script/standard.cpp:330]: Ok.

Comment (bedri) [reply src/script/standard.cpp:330]: > yes, but it is better to add them afterward and if it justifies their inclusion
> if this development works well, in the end there will not be the need to have those very or too specific methods

Ok

File: src/script/interpreter.cpp (hunk 6 above)
Comment (bedri) [reply src/script/interpreter.cpp:973]: I hate spaces :)

**Comment by crypt0p3t3r** [comment]
This PR will be continued at a later stage. Let it be closed for now
EBFs will also give a new perspective on it

**Pull request closed without merging**